### PR TITLE
Working SASL_PLAIN

### PR DESCRIPTION
**Pull request by trthomps**
This solves #119

### CODE REVIEW

**Comment by dasch** [comment]
Thanks for the PR! Sorry for the wait, I've been on vacation this week – will look at this tomorrow.

**Comment by dasch** [comment]
@trthomps can you rebase on the latest master and force push?

File: lib/kafka/sasl_plain_authenticator.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+      unless response.error_code == 0 && response.enabled_mechanisms.include?(PLAIN_IDENT)
+        raise Kafka::Error, "#{PLAIN_IDENT} is not supported."
+      end
+
+      # SASL PLAIN
+      msg = [@authzid.to_s,
+             @username.to_s,
+             @password.to_s].join("\000").force_encoding("utf-8")
+      @encoder.write([msg.bytesize].pack("l>")+msg)
```
Comment (dasch) [inline lib/kafka/sasl_plain_authenticator.rb]: Would `@encoder.write_bytes(msg)` not work here?

File: lib/kafka/sasl_plain_authenticator.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      begin
+        msg = @decoder.bytes
+        raise Kafka::Error, "SASL PLAIN authentication failed." unless msg
+      rescue Errno::ETIMEDOUT
+        raise Kafka::Error, "SASL PLAIN authentication failed." unless msg
```
Comment (dasch) [inline lib/kafka/sasl_plain_authenticator.rb]: The last error message should probably include the reason why it failed, e.g.

> SASL PLAIN authentication timed out

File: lib/kafka/sasl_plain_authenticator.rb (hunk 1 above)
Comment (trthomps) [inline lib/kafka/sasl_plain_authenticator.rb]: I had that but for whatever reason it wouldn't work.

File: lib/kafka/sasl_plain_authenticator.rb (hunk 2 above)
Comment (trthomps) [inline lib/kafka/sasl_plain_authenticator.rb]: Good call.

File: lib/kafka/sasl_plain_authenticator.rb (hunk 1 above)
Comment (dasch) [inline lib/kafka/sasl_plain_authenticator.rb]: `#write_bytes` should do exactly what you're doing here (https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/protocol/encoder.rb#L90), with the only difference being when `msg` is `nil` – which it definitely shouldn't be here. Can you try again?

File: lib/kafka/sasl_plain_authenticator.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
+module Kafka
+  class SaslPlainAuthenticator
```
Comment (dasch) [inline lib/kafka/sasl_plain_authenticator.rb]: I'd like to see a test that exercises this code, maybe against a fake SASL server implementation. Currently none of this code is run during testing.

Comment (trthomps) [inline lib/kafka/sasl_plain_authenticator.rb]: Agreed.  I broke the `FakeServer` class out and added support for SASL PLAIN authentication.

**Comment by trthomps** [comment]
It doesn't look like the failed functional tests are related to this PR.

**Comment by tmeinlschmidt** [comment]
@trthomps good job!

**Pull request merged**